File: app/src/test/java/com/dominikgaller/alpinebooking/booking/inbound/driver/ConfirmTourBookingDriverTest.java
package com.dominikgaller.alpinebooking.booking.inbound.driver;

import com.dominikgaller.alpinebooking.booking.core.domain.AvailableCapacity;
import com.dominikgaller.alpinebooking.booking.core.domain.BookingId;
import com.dominikgaller.alpinebooking.booking.core.domain.ParticipantContact;
import com.dominikgaller.alpinebooking.booking.core.domain.ParticipantCount;
import com.dominikgaller.alpinebooking.booking.core.domain.TourBooking;
import com.dominikgaller.alpinebooking.booking.core.domain.TourDate;
import com.dominikgaller.alpinebooking.booking.core.domain.TourId;
import com.dominikgaller.alpinebooking.shared.domain.event.DomainEvent;
import com.dominikgaller.alpinebooking.booking.core.domain.event.TourBookingConfirmed;
import com.dominikgaller.alpinebooking.booking.core.domain.exception.BookingNotFoundException;
import com.dominikgaller.alpinebooking.booking.core.domain.exception.InvalidBookingStateException;
import com.dominikgaller.alpinebooking.booking.core.inport.command.ConfirmTourBookingCommand;
import com.dominikgaller.alpinebooking.booking.core.inport.result.ConfirmTourBookingResult;
import com.dominikgaller.alpinebooking.booking.core.outport.ClockPort;
import com.dominikgaller.alpinebooking.booking.core.outport.DomainEventPublisher;
import com.dominikgaller.alpinebooking.booking.core.outport.TourBookingRepository;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

/**
 * Spring-free unit tests for {@link ConfirmTourBookingDriver}.
 *
 * <p>All outport dependencies are replaced by in-line stub implementations.
 * No Spring context is loaded.
 */
class ConfirmTourBookingDriverTest {

    private static final Instant NOW = Instant.parse("2026-03-03T12:00:00Z");
    private static final LocalDate FUTURE_DATE = LocalDate.of(2026, 6, 15);

    private final BookingStubRepository repository = new BookingStubRepository();
    private final PublishCapturingPublisher publisher = new PublishCapturingPublisher();
    private final FixedClockPort clock = new FixedClockPort(NOW);

    private final ConfirmTourBookingDriver driver =
            new ConfirmTourBookingDriver(repository, publisher, clock);

    private TourBooking requestedBooking() {
        final BookingId id = BookingId.generate();
        final TourBooking booking = TourBooking.request(
                id,
                new TourId("TOUR-42"),
                new TourDate(FUTURE_DATE),
                new ParticipantCount(3),
                new AvailableCapacity(10),
                new ParticipantContact("Alice", "alice@example.com"),
                NOW);
        booking.pullDomainEvents(); // drain TourBookingRequested — simulates reconstitution from DB
        return booking;
    }

    // ── Happy path ───────────────────────────────────────────────────────────

    @Test
    void confirm_happyPath_returnsConfirmedStatus() {
        final TourBooking booking = requestedBooking();
        repository.preload(booking);

        final ConfirmTourBookingResult result =
                driver.confirm(new ConfirmTourBookingCommand(booking.bookingId().value().toString()));

        assertThat(result.status()).isEqualTo("CONFIRMED");
    }

    @Test
    void confirm_happyPath_callsUpdateOnRepository() {
        final TourBooking booking = requestedBooking();
        repository.preload(booking);

        driver.confirm(new ConfirmTourBookingCommand(booking.bookingId().value().toString()));

        assertThat(repository.updatedBookings()).hasSize(1);
        assertThat(repository.updatedBookings().get(0).status().name()).isEqualTo("CONFIRMED");
    }

    @Test
    void confirm_happyPath_publishesTourBookingConfirmedEvent() {
        final TourBooking booking = requestedBooking();
        repository.preload(booking);

        driver.confirm(new ConfirmTourBookingCommand(booking.bookingId().value().toString()));

        assertThat(publisher.publishedEvents()).hasSize(1);
        assertThat(publisher.publishedEvents().get(0)).isInstanceOf(TourBookingConfirmed.class);
    }

    // ── Not found ────────────────────────────────────────────────────────────

    @Test
    void confirm_throwsBookingNotFoundException_whenNotFound() {
        final String unknownId = BookingId.generate().value().toString();

        assertThatExceptionOfType(BookingNotFoundException.class)
                .isThrownBy(() -> driver.confirm(new ConfirmTourBookingCommand(unknownId)));
    }

    // ── Invalid state ────────────────────────────────────────────────────────

    @Test
    void confirm_throwsInvalidBookingStateException_whenAlreadyConfirmed() {
        final TourBooking booking = requestedBooking();
        booking.confirm(NOW); // transition to CONFIRMED
        booking.pullDomainEvents(); // drain events so the stub starts clean
        repository.preload(booking);

        assertThatExceptionOfType(InvalidBookingStateException.class)
                .isThrownBy(() -> driver.confirm(
                        new ConfirmTourBookingCommand(booking.bookingId().value().toString())));
    }

    @Test
    void confirm_doesNotCallUpdate_whenStateInvalid() {
        final TourBooking booking = requestedBooking();
        booking.confirm(NOW);
        booking.pullDomainEvents();
        repository.preload(booking);

        assertThatExceptionOfType(InvalidBookingStateException.class)
                .isThrownBy(() -> driver.confirm(
                        new ConfirmTourBookingCommand(booking.bookingId().value().toString())));

        assertThat(repository.updatedBookings()).isEmpty();
    }

    // ── Stub implementations ─────────────────────────────────────────────────

    private static class BookingStubRepository implements TourBookingRepository {
        private final Map<String, TourBooking> store = new LinkedHashMap<>();
        private final List<TourBooking> updated = new ArrayList<>();

        void preload(final TourBooking booking) {
            store.put(booking.bookingId().value().toString(), booking);
        }

        @Override
        public void save(final TourBooking booking) {
            store.put(booking.bookingId().value().toString(), booking);
        }

        @Override
        public Optional<TourBooking> findById(final BookingId bookingId) {
            return Optional.ofNullable(store.get(bookingId.value().toString()));
        }

        @Override
        public void update(final TourBooking booking) {
            updated.add(booking);
        }

        List<TourBooking> updatedBookings() {
            return Collections.unmodifiableList(updated);
        }
    }

    private static class PublishCapturingPublisher implements DomainEventPublisher {
        private final List<DomainEvent> events = new ArrayList<>();

        @Override
        public void publish(final DomainEvent event) {
            events.add(event);
        }

        List<DomainEvent> publishedEvents() {
            return Collections.unmodifiableList(events);
        }
    }

    private record FixedClockPort(Instant instant) implements ClockPort {
        @Override
        public Instant now() {
            return instant;
        }
    }
}

File: src/test/kotlin/com/example/contractmanagement/booking/inbound/driver/ConfirmTourBookingDriverTest.kt
package com.example.contractmanagement.booking.inbound.driver

import com.example.contractmanagement.booking.core.domain.tourbooking.AvailableCapacity
import com.example.contractmanagement.booking.core.domain.tourbooking.BookingId
import com.example.contractmanagement.booking.core.domain.tourbooking.ParticipantContact
import com.example.contractmanagement.booking.core.domain.tourbooking.ParticipantCount
import com.example.contractmanagement.booking.core.domain.tourbooking.TourBooking
import com.example.contractmanagement.booking.core.domain.tourbooking.TourBookingStatus
import com.example.contractmanagement.booking.core.domain.tourbooking.TourDate
import com.example.contractmanagement.booking.core.domain.tourbooking.event.TourBookingConfirmed
import com.example.contractmanagement.booking.core.domain.tourbooking.exception.BookingNotFoundException
import com.example.contractmanagement.booking.core.domain.tourbooking.exception.InvalidBookingStateException
import com.example.contractmanagement.booking.core.inport.command.ConfirmTourBookingCommand
import com.example.contractmanagement.booking.core.outport.TourBookingRepository
import com.example.contractmanagement.shared.domain.TourId
import com.example.contractmanagement.shared.outport.ClockPort
import com.example.contractmanagement.shared.outport.DomainEventPublisher
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.check
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import java.time.Instant
import java.time.LocalDate
import java.util.UUID

/**
 * Use case tests for [ConfirmTourBookingDriver] (UC02).
 *
 * SDD: see `documentation/use-cases/uc02-confirm-tour-booking.spec.md`.
 */
class ConfirmTourBookingDriverTest {
    private val now: Instant = Instant.parse("2026-06-01T10:00:00Z")
    private val uuid: UUID = UUID.fromString("11111111-1111-1111-1111-111111111111")

    private val repository: TourBookingRepository = mock()
    private val eventPublisher: DomainEventPublisher = mock()
    private val clockPort: ClockPort = mock { on { now() } doReturn now }

    private val driver = ConfirmTourBookingDriver(repository, eventPublisher, clockPort)
    private val command = ConfirmTourBookingCommand(uuid.toString())

    private fun bookingIn(status: TourBookingStatus): TourBooking =
        TourBooking.reconstitute(
            bookingId = BookingId.of(uuid),
            tourId = TourId("tour-42"),
            tourDate = TourDate(LocalDate.parse("2026-07-01")),
            participantCount = ParticipantCount(2),
            availableCapacity = AvailableCapacity(10),
            contact = ParticipantContact("Ada Lovelace", "ada@example.com"),
            status = status,
        )

    @Test
    fun confirm_returnsConfirmedStatus() {
        whenever(repository.findById(any())).thenReturn(bookingIn(TourBookingStatus.REQUESTED))

        assertThat(driver.confirm(command).status).isEqualTo(TourBookingStatus.CONFIRMED.name)
    }

    @Test
    fun confirm_updatesTheAggregate() {
        whenever(repository.findById(any())).thenReturn(bookingIn(TourBookingStatus.REQUESTED))

        driver.confirm(command)

        verify(repository).update(check { assertThat(it.status).isEqualTo(TourBookingStatus.CONFIRMED) })
    }

    @Test
    fun confirm_usesClockPort_forTheEventTimestamp() {
        whenever(repository.findById(any())).thenReturn(bookingIn(TourBookingStatus.REQUESTED))

        driver.confirm(command)

        verify(eventPublisher).publish(
            check<TourBookingConfirmed> { assertThat(it.occurredAt).isEqualTo(now) },
        )
    }

    @Test
    fun confirm_throwsBookingNotFoundException_whenNoBookingHasThatIdentity() {
        whenever(repository.findById(any())).thenReturn(null)

        assertThatThrownBy { driver.confirm(command) }
            .isInstanceOf(BookingNotFoundException::class.java)
    }

    @Test
    fun confirm_throwsIllegalArgumentException_whenBookingIdIsMalformed() {
        // The source of the IllegalArgumentException -> 400 backstop: parsing an identifier
        // out of a path variable. Asserted here as well as at the boundary, because the
        // exception has to be thrown before it can be mapped, and it once surfaced as a 500.
        assertThatThrownBy { driver.confirm(ConfirmTourBookingCommand("not-a-uuid")) }
            .isInstanceOf(IllegalArgumentException::class.java)

        verify(repository, never()).update(any())
    }

    @Test
    fun confirm_propagatesInvalidBookingStateException_andUpdatesNothing() {
        whenever(repository.findById(any())).thenReturn(bookingIn(TourBookingStatus.ACTIVE))

        assertThatThrownBy { driver.confirm(command) }
            .isInstanceOf(InvalidBookingStateException::class.java)

        verify(repository, never()).update(any())
        verify(eventPublisher, never()).publish(any())
    }
}
